design/tricky-choices/dag-pb-forms-impl-and-use.md

```diff
@@ -0,0 +1,126 @@
+
+Rather than using codec-level JSONificiation, js-ipfs uses [custom code _within_ `dag get`](https://github.com/ipfs/js-ipfs/blob/ipfs@0.55.3/packages/ipfs-cli/src/commands/dag/get.js#L82-L89) to print the same forms that go-ipfs provides, translating `{Data,Links:[{Name,Tsize,Hash},...]}` to `{"data":"..","links":[{"Name":"..","Tsize":x,"Hash":{"/":"Qm..."}},...]}`. **But it can't do the reverse**, if you `dag put` then it _must_ be in the raw codec form of `{Data,Links:[{Name,Tsize,Hash},...]}` (but JSON). If you feed `dag get` dag-pb output back into `dag put` then you'll end up with an empty node because it (1) won't find data in the JSON because of casing, and (2) won't error that you have extraneous properties.
+
+During the development of the JavaScript stack, in 2017, there was a conscious decision to [not “translate”](https://github.com/ipld/js-ipld-dag-pb/pull/29) the forms that arise out of the data and just present them as they are (where "as they are" really means, "as the PB schema says they are", this is the crux of the topic of this post). So we get objects with `Data` and `Links`, even though the original code had `data` and `links`, and links have the `Tsize` and `Hash` properties, translating the protobuf schema.
+
+## DAG-PB specification
+
+Prior to 2020, [the dag-pb specification](https://ipld.io/specs/codecs/dag-pb/spec/) only referenced the protobuf schema and included notes about ordering and the some details about the differences in pathing between the two stacks.
+
+In https://github.com/ipld/specs/issues/245 and then https://github.com/ipld/specs/pull/247 an IPLD Schema form of dag-pb was merged. There was some discussion about casing of names at https://github.com/ipld/specs/pull/247#discussion_r390220519 where it was resolved that we should match the protobuf schema. We were apparently oblivious to the additional `ProtoNode` representation, _or_ we didn't consider it relevant to the "data model" form of dag-pb. (There is probably some valid critique of the previous siloing between IPFS and IPLD here too since the IPLD team haven't noticed any of these problems until now.)
+
+The IPLD Schema has since been used to rewrite the new JS and Go codecs for dag-pb, but this was not a major departure since these codecs (originally js-ipld-dag-pb and go-merkledag/pb) at their protobuf level were always using these forms. We were adding additiona lenses onto these codecs before they presented themselves at the IPFS `dag` API - via `ProtoNode`, `ipld.Link` and the custom printing forms in js-ipfs `dag get`.
```

Comment:
```suggestion
The IPLD Schema has since been used to rewrite the new JS and Go codecs for dag-pb, but this was not a major departure since these codecs (originally js-ipld-dag-pb and go-merkledag/pb) at their protobuf level were always using these forms. We were adding additional lenses onto these codecs before they presented themselves at the IPFS `dag` API - via `ProtoNode`, `ipld.Link` and the custom printing forms in js-ipfs `dag get`.
```